finality-aleph/src/party/impls.rs

```diff
@@ -0,0 +1,366 @@
+    ) -> Result<(), Self::Error> {
+        let node_id = match self.node_idx(authorities).await {
+            Some(id) => id,
+            None => return Err(SessionManagerError::NotAuthority),
+        };
+        let authority_verifier = AuthorityVerifier::new(authorities.to_vec());
+        let authority_pen =
+            AuthorityPen::new(authorities[node_id.0].clone(), self.keystore.clone())
+                .await
+                .expect("The keys should sign successfully");
+        self.session_manager
+            .early_start_validator_session(session, authority_verifier, node_id, authority_pen)
+            .map_err(SessionManagerError::ManagerError)
+    }
+
+    fn start_nonvalidator_session(
+        &self,
+        session: SessionId,
+        authorities: &[AuthorityId],
+    ) -> Result<(), Self::Error> {
+        let authority_verifier = AuthorityVerifier::new(authorities.to_vec());
+
+        self.session_manager
+            .start_nonvalidator_session(session, authority_verifier)
+            .map_err(SessionManagerError::ManagerError)
+    }
+
+    fn stop_session(&self, session: SessionId) -> Result<(), Self::Error> {
+        self.session_manager
+            .stop_session(session)
+            .map_err(SessionManagerError::ManagerError)
+    }
+
+    async fn node_idx(&self, authorities: &[AuthorityId]) -> Option<NodeIndex> {
+        let our_consensus_keys: HashSet<_> = self
+            .keystore
+            .keys(KEY_TYPE)
+            .await
+            .unwrap()
+            .into_iter()
+            .collect();
+        trace!(target: "aleph-data-store", "Found {:?} consensus keys in our local keystore {:?}", our_consensus_keys.len(), our_consensus_keys);
+        authorities
+            .iter()
+            .position(|pkey| our_consensus_keys.contains(&pkey.into()))
+            .map(|id| id.into())
+    }
+}
+
+pub struct ChainStateImpl<B, BE, CFA>
+where
+    B: BlockT,
+    BE: Backend<B>,
+    CFA: ClientForAleph<B, BE>,
+{
+    pub client: Arc<CFA>,
+    pub _phantom: PhantomData<(B, BE)>,
+}
+
+impl<B, BE, CFA> ChainState<B> for ChainStateImpl<B, BE, CFA>
+where
+    B: BlockT,
+    BE: Backend<B>,
+    CFA: ClientForAleph<B, BE>,
+{
+    fn best_block_number(&self) -> <B as Block>::Number {
+        self.client.info().best_number
+    }
+    fn finalized_number(&self) -> <B as Block>::Number {
+        self.client.info().finalized_number
+    }
+}
+
+pub struct SessionInfoImpl {
+    session_period: SessionPeriod,
+}
+
+impl SessionInfoImpl {
+    pub fn new(session_period: SessionPeriod) -> Self {
+        Self { session_period }
+    }
+}
+
+impl<B: BlockT> SessionInfo<B> for SessionInfoImpl {
+    fn session_id_from_block_num(&self, n: NumberFor<B>) -> SessionId {
+        SessionId(n.saturated_into::<u32>() / self.session_period.0)
+    }
+
+    fn last_block_of_session(&self, session_id: SessionId) -> NumberFor<B> {
+        ((session_id.0 + 1) * self.session_period.0 - 1).into()
```

Comment:
yes, it is kinda cumbersome to remove as it involves other files not related to this pr. I could do it in this pr if you think it will make more sense though?

Reply:
I'm okay with doing it later.